Creating a preview of the outline proposed

kinow opened the pull request:
Closes #222 
Closes #228 
Closes #157 (invalid/outdated)
Closes #141 (I think)
Closes #64 (in prerequisites)
Closes #6 (I think? Then we can open new PR's or issues for improving the FAQ if needed?)
Closes #2 (I think? Maybe invalid/outdated now with this change, and also with the novice tutorial work?)
Closes #180 (invalid/outdated the missing chapter won't be a problem as we don't have numbers in chapters - although we do have numbers in sections now… but that issue at least is fixed, and it really bothered me too :smile: )
Closes #217 (link added to the first paragraph of the documentation)
Closes #233 (new `.github/workflows/documentation-links.yaml` file added as per their docs, also includes link to their docs)

A draft PR with a preview of #222. Following @tetron's suggestion so that it's easier for others to review & send feedback, and also allows me to try to write some of the new text (or copy existing text into new locations, add redirects for old pages, etc.).

- [x] Reproduce the proposed outline with Sphinx
- [x] Add content for the first sections:
  - [x] Landing Page
  - [x] Quick Start
  - [x] Prerequisites
  - [x] Basic Concepts
- [x] Build the old pages so there are no/few links broken after the upgrade (orphan pages in Sphinx)
- [x] Add a note to old pages about the new content
  - [x] Episode 1
  - [x] Episode 2
  - [x] Episode 3
  - [x] Episode 4
  - [x] Episode 5
  - [x] Episode 6
  - [x] Episode 7
  - [x] Episode 8
  - [x] Episode 9
  - [x] Episode 10
  - [x] Episode 11
  - [x] Episode 12
  - [x] Episode 13
  - [x] Episode 14
  - [x] Episode 15
  - [x] Episode 16
  - [x] Episode 17
  - [x] Episode 19
  - [x] Episode 20
  - [x] Episode 21
  - [x] Episode 22
  - [x] Episode 23
  - [x] Episode 24
  - [x] Setup

Once the items above are complete (or at least the Basic Concepts page as well) it will be ready for an initial review from @mr-c , @tetron , @swzCuroverse , and others :-) Then what'll be left are the Core Concepts, FAQ, Tutorials, and How-Tos, which I think will be simpler, but may take longer as some require writing new content.

kinow commented:
I think I found a way to deploy it to my `gh-pages` without touching the `common-workflow-language/user_guide` branch:

```bash
# build passing the `html_baseurl` value
rm -rf _build && sphinx-build -b html . _build/html/ -Dhtml_baseurl="/user_guide/" -W
# use `ghp-import` to push to `gh-pages` in my fork called `kinow`
ghp-import -p -r kinow -b gh-pages ./_build/html/
```

It is not rendering 100% correct for me, but I think that might be due to cache on the GH Pages server side. Looking at my `gh-pages` branch everything looks OK. Will try again later :wave: 

https://kinow.github.io/user_guide/index.html

mr-c commented:
There is also the RTD PR preview: https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/

mr-c commented:
There is also the RTD PR preview: https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/

mr-c commented:
Seems like this would break all existing links to specific pages in the user guide. We can re-organize, but we should add redirects or other text to hidden pages so that old links aren't completely useless.

kinow commented:
> Seems like this would break all existing links to specific pages in the user guide. We can re-organize, but we should add redirects or other text to hidden pages so that old links aren't completely useless.

My idea is to slowly move the content from the episodes into the new sections of the user guide. Once the content has been moved, or is not necessary any longer, the text could be replaced by some text like:

“The content of this page has been moved to Section A, and Section B. If you would like to access the old content, you can view the source on [GitHub][link-to-old-page-commit-with-content], or to visualize it in the [Internet Archive WayBack Machine][link-to-wayback-machine].”

Ideally, I think it would be nice to also send a [410](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/410), permanently gone status with the response, so search engines know the page is not to be considered in the search results. Or maybe a 301 or 30x if it has a single page to redirect. But that's probably hard with GH pages (probably the closest are the meta refreshes or JS navigation changes).

swzCuroverse commented:
Hi Bruno -

This is looking really nice.  Thank you!

Here are some thoughts.  (I will do a different message for each section)

* In Quick Start Section:
** maybe consider explaining what an implementation is - and having a note that links to the list on the website of the current implementations
**  I am not sure why you use the term [JOB_ORDER] for the input file.  I have never heard it called that before, but it might be that way in material I haven't come across. 
** I personally find the name a little confusing so I would probably call it [INPUTS_OBJECT] based off of the CWL help file.  I would also mention that it can be YML/YAML or JSON and mention why you might one over the other.  I tend to always use YAML but I am not sure about others, so I don't know if the example case should be in YAML or JSON.

swzCuroverse commented:
* In Prerequisites sections:
** CWL implementation --> again, maybe we should explain what it is.  I think people find this confusing -- and don't know there are other implementations out there and/or why you would use them
** Is this needed? #!/usr/bin/env cwl-runner
** I would explain this in the text instead of the cwl -- also there is typo (extra c)
  # `true` is a Linux command that exits with exit code `0` (succcess).
** From the output, I find it hard to parse the command you run.  I wonder if you want to put the command you will run in the text as well.
** Maybe a little explanation of what validation does other than checks that it is valid?
** AH - here is the cwl runner part -- I would maybe just make it another cwl file for clarity so that there is as little extra on the intro files as needed. 
** Oh, I see the implementation list at the bottom.  Might be good to explain why you might use another implementation before the list.

kinow commented:
Hi @swzCuroverse !

Thanks a lot for the great feedback.

>In Quick Start Section:
>** maybe consider explaining what an implementation is - and having a note that links to the list on the website of the current implementations

I've replaced that by CWL Runners. There's still a link to CWL Implementations at the very bottom of the page, but hopefully that won't distract readers.

>** I am not sure why you use the term [JOB_ORDER] for the input file. I have never heard it called that before, but it might be that way in material I haven't come across.
>** I personally find the name a little confusing so I would probably call it [INPUTS_OBJECT] based off of the CWL help file. I would also mention that it can be YML/YAML or JSON and mention why you might one over the other. I tend to always use YAML but I am not sure about others, so I don't know if the example case should be in YAML or JSON.

Glad I'm not the only one! I mentioned it in #222, it took me a while to figure out what was a job order. I had seen it in the `cwltool` help and in other documents and also being mentioned in issues & discussions. Eventually it clicked and I realized it was the file with the input parameters for the workflow.

I thought it would be good to introduce this term earlier to users, but actually something like Inputs Object would be a lot more intuitive. @tetron , @mr-c maybe we could phase-out the term Job Order, and slowly replace it in places like cli help, user guide, etc?

I've replaced it in this PR by Inputs Object, and also added the note about it being JSON or YAML. Thanks!

>In Prerequisites sections:
>** CWL implementation --> again, maybe we should explain what it is. I think people find this confusing -- and don't know there are other implementations out there and/or why you would use them

The Quick Start is using CWL Runner, and the user must be first introduced to the term now here in Prerequisites section. I've added the same link to the implementations at the bottom, and also at a note right after where it appears, alongside a short explanation (will self-edit what I wrote later today).

> ** Is this needed? #!/usr/bin/env cwl-runner
** AH - here is the cwl runner part -- I would maybe just make it another cwl file for clarity so that there is as little extra on the intro files as needed.

Good point. I tried to not assume the reader knew what a shebang was, but maybe introducing it before the section about the `cwl-runner` command where it's used could confuse readers. 

> ** I would explain this in the text instead of the cwl -- also there is typo (extra c)
> ** true is a Linux command that exits with exit code 0 (succcess).

Typo fixed! Thanks!

>** From the output, I find it hard to parse the command you run. I wonder if you want to put the command you will run in the text as well.

Me too!!! I spent almost 30 minutes fiddling with the Sphinx settings, trying to change the command output (syntax highlighting and formatting&layout), but I decided to keep writing and leave that for later.

I will have another go at improving it later. I also want to make the output less platform & user-dependent. At the moment it includes my user name, home directory, etc. Since Sphinx uses Python, we can probably write something (or use some existing extension) that runs those commands in a container—for instance—to get the same output independent of the user or environment where the documentation it being built on.

Thank you!!! :rocket:

kinow commented:
Now for tomorrow: start writing the basic concepts section, where I will try to use some images too, like diagrams to illustrate the basic concepts for readers :nerd_face:

kinow commented:
Added the first graph. It was created with Graphviz, following what was used in the [Cylc doc](https://github.com/cylc/cylc-doc/blob/57a6ba35fc89be2f6037cf30f88fed6e377bdc77/src/tutorial/introduction.rst) repository (credit to them, I learned about it there :grimacing: ).

The first added contains the processing units of the CWL object model. It uses `cwl-utils` to find the `Process` class, then calls `__subclasses__()` and creates the graphviz plot. Sphinx has a built-in extension for rendering these diagrams.

In theory, when we have a new specification, it should either render the updated object model, or require little maintenance to update the graph :crossed_fingers: 

![image](https://user-images.githubusercontent.com/304786/180709632-8a81bd3f-6aab-4c49-8270-8c183cbcde2e.png)

mr-c commented:
> I'm fixing the Travis build, but I think the RTD would need to be updated to install the dependencies.


https://docs.readthedocs.io/page/config-file/v2.html
suggest adding a `.readthedocs.yaml` with the following (or similar)
```
version: 2

mkdocs:
  configuration: null

python:
  install:
    - method: pip
      path: .
      extra_requirements:
        - all
```

kinow commented:
That fixed it @mr-c ! Thanks!!!

kinow commented:
To have Sphinx to build a page, without adding it to a `toctree`, we just need to make it an orphan page :slightly_smiling_face: 

Example: `file.md`:

```
---
orphan: true
---

# Heading 1

....
```

kinow commented:
Pushing a new commit with the old episodes as orphan pages, and will slowly start adding notes in each page telling users the page is out-of-date, and pointing to a new page when we have a new page with the same content.

![image](https://user-images.githubusercontent.com/304786/181393372-d0939bc7-adab-4ff9-8495-2ce97861e2b3.png)

kinow commented:
Going to take a break over the weekend, then self-edit what I wrote today on Monday and try to add more notes to the chapters of the old user guide :wave:

tetron commented:
* Have Sphinx auto-number the chapters and sections
* GraphViz figures in Basic Concepts not rendering
* "Tutorials" and "YAML Guide" should be closer to the top and/or part of Quick Start
* "Miscellaneous" should be part of "How To" or "FAQ"

tetron commented:
Outgoing links should go to the rendered version, not github:

https://doc.arvados.org/user/cwl/arvados-vscode-training.html

kinow commented:
Thanks @tetron !!!

>GraphViz figures in Basic Concepts not rendering

Good catch! I logged-in to ReadTheDocs to check the logs:

```
WARNING: dot command 'dot' cannot be run (needed for graphviz output), check the graphviz_dot setting
```

![image](https://user-images.githubusercontent.com/304786/182480075-ea071008-9fcb-4cbf-9dac-e10b9602e25e.png)

How it was rendering before:

![image](https://user-images.githubusercontent.com/304786/182481066-88841abf-7c4e-4397-980f-3bb65285599c.png)

After a hard-refresh (CTRL + SHIFT + R on Linux/Firefox):

![image](https://user-images.githubusercontent.com/304786/182481137-e4020569-56c8-4513-b800-173ff1e2102a.png)

This issue should be fixed now. Thanks!

kinow commented:
>Have Sphinx auto-number the chapters and sections

I think the numbered table of content requires the top menu to be reduced. The numbers added cause it to wrap the menu items. Would be nice if we could have numbers only at the left sidebar and in the landing page. Will check later if that's doable. But WDYT of this preview?

![image](https://user-images.githubusercontent.com/304786/182500609-64d84377-0673-4b73-803f-a2a95b0bcf90.png)

kinow commented:
@mr-c , @tetron this PR is now ready for a pre-review :grimacing: 

The new outline was created with the following in mind:

- Quick Start
  - I copied this idea from other user guides where they have something really brief about the tool/technology. The target audience for this section are pretty much everybody. A manager who never heard about CWL going to a meeting that has CWL in the agenda, a new CWL intern, or a workflow engineer coming from another technology/standard.
  - I tried to make it as simple as possible, and avoided adding too many links to other parts of the guide (the natural flow should be to progress to Next Page)
- Prerequisites
  - Same as the `setup.md` page we had. Focused on CWL + `cwltool`, but without showing all possible ways to install `cwltool`. Or maybe we should list conda/pip/etc? I thought about showing `pip`, mentioning other ways and linking to the `cwltool` page.
  - This page is targeted to a user who wants to try the examples in the user guide.
- Basic Concepts
  - This is the page I wish I had before getting started with CWL & `cwltool`. Something that introduces me to the jargon used in CWL, and help me preparing to the rest of the user guide.
  - The target audience are workflow engineers, Outreachy/GSoC interns, researchers, bioinformaticians.
- Core Concepts
  - Here's what you need to read, understand, and maybe write some simple workflows. I mainly used existing content for the pages in this section.
  - Some pages are missing content, but I prefer to have a review before starting writing.
  - The target audience will be for workflow engineers, interns, researchers, bioinformaticians.
- Advanced Topics
  - It was under Core Concepts, but I thought it would make more sense to move it to its own section.
  - It will contain more advanced topics like staging files, containers, advanced / difficult input/output parameter handling, debug JS, typescript, maybe schema-salad, etc.
  - The target audience are workflow engineers, people responsible for troubleshooting workflows, admins deploying CWL runners in production.
- HowTos/Tutorials/FAQ - as the name implies, although I just converted the existing TODO's into content, more to have an idea of what the layout would look like. We still need to think what we want in each of these sections, and if we want to keep/modify/drop them.

Other notes:

- The existing links must still work in the preview build.
- I've added Graphviz graphs, with a Sphinx extension, but we can use another tool if needed.
- Would be great to have a review of: the new Outline, the Landing Page, the Prerequisites, the Basic Concepts. The other sections are still WIP, and I didn't spend much on them.

kinow commented:
I've changed the titles of the old pages to "This page has moved", removed the contents, and added the [`sphinx-reredirect`](https://documatt.gitlab.io/sphinx-reredirects/usage.html) extension.

If you navigate to `01-introduction` or to `01-introduction/index.html`, both URL's are now redirected to the `quick-start.html` URL. The redirects are defined in `conf.py`. Some links point to anchors like `workflows.html#nested-workflows`.

The user probably won't even see the page title or the message that it has moved. I disabled JS in NoScript to check if I could see the old page, but the HTML meta refresh was still fast enough to redirect the user :)

kinow commented:
>"Tutorials" and "YAML Guide" should be closer to the top and/or part of Quick Start

If we move these up, it means that from quick-start or basic concepts, users navigating with the "Next" link at the end will go from the quick start to the tutorials, before looking at the basic concepts. The pages are self-contained, but the order was intentional so that users could go from landing → quick-start → prerequisites → basic concepts → core concepts → advanced. Then after that I don't have a preferred order, and now we have external links with HowTos/Tutorials and FAQ/Extras. WDYT?

kinow commented:
>"Miscellaneous" should be part of "How To" or "FAQ"

Done! Looks better! I think the Extras having just the YAML now is a bit weird. But that should be fine if we will add more to the Extras. Otherwise maybe we could find a new home for the YAML guide (or link to an external resource if any contains the same info).

kinow commented:
>Outgoing links should go to the rendered version, not github:

Done! Thanks Peter!

kinow commented:
In the last commits I've:

- fixed links in the sections migrated from the old user guide; some links were pointing to 00-chapter-name and now are pointing to the new links in core-concepts/chapter-name, etc.
- hidden the Operations and Requirements & Hints sections that were blank still pending content (the pages still exist, but orphaned)
- added a small paragraph to core concepts and to advanced topics sections

I think the current version, if approved, could already be merged. It contains the same content as the old user guide, plus the new content. We still need to work on @tetron 's feedback about numbered sections and moving sections, and I believe @mr-c will have other feedback.

I'll ask the APAC community to start having a look at the preview too, to see if they have other ideas.

Thank you!

kinow commented:
Added the text for the Operation. We discussed it in the APAC meeting last Friday, and after the meeting I did some experiments and shared in the workflow-jp-meetup slack. I used these examples, and the text from `operations.yml` from the v1.2.1 to write this text :+1: 

I've also linked this new page from the `basic-concepts.md`.

kinow commented:
Reviewed all the code blocks, using the updated syntax for Sphinx. Also updated the caption/legend of code blocks. Before we had the caption in italic, as a paragraph before the image. This is now handled by sphinx, since the directives support a caption attribute. It also handles creating links to the elements with captions, which means users can share the code blocks directly in issues/email/etc.

Before: 

![image](https://user-images.githubusercontent.com/304786/183562356-93f3059d-6a06-4425-abdd-4df9b5c9a762.png)

Now (note the pink/red text above the code block):

![image](https://user-images.githubusercontent.com/304786/183562878-9abca26c-c75f-4eb6-a22b-ebed655ddfe1.png)

kinow commented:
Modified the best practices formatting. Now it displays a list instead of paragraphs prefixed by the square/check mark.

tetron commented:
Process discussion -- should define a process as a thing that takes inputs, produces outputs, and its behavior can be affected by the inputs, requirements and hints.  Should also describe requirements and hints and how they relate to the CWL extension mechanism.

Then command line, workflow and expression tool can just be a paragraph or two describing how they concretely implement the concept of a process specifically by running a program, or orchestrating a workflow, or running a bit of javascript code.

tetron commented:
Suggest "Core Concepts" -> "Essential Topics"

tetron commented:
Suggest moving the YAML guide to the essential topics section, with a note about if you already YAML you can skip it.

tetron commented:
FAQ and HOW-TO should be merged

tetron commented:
If HOW-TO is going to still be one big page, it should start with a table of contents so the user can quickly find what they are looking for -- alternately, each topic should be listed as a 2nd level under the HOW-TO on the left.  The right hand table of contents isn't sufficient because it disappears if the window is too narrow.

tetron commented:
Sections under "Introduction" should be separate pages

kinow commented:
>Suggest "Core Concepts" -> "Essential Topics"

Done!

>Suggest moving the YAML guide to the essential topics section, with a note about if you already YAML you can skip it.

Done!

![image](https://user-images.githubusercontent.com/304786/183778290-902f5dad-8971-4c94-aa2b-ce2a42449f9c.png)

>Sections under "Introduction" should be separate pages

Done!

kinow commented:
>FAQ and HOW-TO should be merged

Done!

>If HOW-TO is going to still be one big page, it should start with a table of contents so the user can quickly find what they are looking for -- alternately, each topic should be listed as a 2nd level under the HOW-TO on the left. The right hand table of contents isn't sufficient because it disappears if the window is too narrow.

I would like to try a single page first. It's easier to manage, and—like you said—in case we find any entry that is too long, it's probably a good indicator that it deserves its own place somewhere in the user guide.

Done the table of contents with the Sphinx directive - https://docutils.sourceforge.io/docs/ref/rst/directives.html#table-of-contents

![image](https://user-images.githubusercontent.com/304786/183785672-fb3eda05-13be-4f36-9e8f-b9c3cf7f36ae.png)

kinow commented:
>Process discussion -- should define a process as a thing that takes inputs, produces outputs, and its behavior can be affected by the inputs, requirements and hints. Should also describe requirements and hints and how they relate to the CWL extension mechanism.
>
>Then command line, workflow and expression tool can just be a paragraph or two describing how they concretely implement the concept of a process specifically by running a program, or orchestrating a workflow, or running a bit of javascript code.

I will start thinking about this one today, then write it tomorrow. I think this is the last bit missing from @tetron 's feedback. I think I also addressed @mr-c 's feedback from last week. Marking it ready for review. Just pending the feedback from the comment above.

kinow commented:
>Process discussion -- should define a process as a thing that takes inputs, produces outputs, and its behavior can be affected by the inputs, requirements and hints. Should also describe requirements and hints and how they relate to the CWL extension mechanism.
>
>Then command line, workflow and expression tool can just be a paragraph or two describing how they concretely implement the concept of a process specifically by running a program, or orchestrating a workflow, or running a bit of javascript code.

Had a go at combining the processes sections and the requirement section into a single section. I think I addressed all your feedback @tetron , thanks a lot! :rocket: 

Ready for review @tetron , @mr-c :bow:

kinow commented:
For reviewers:

- This PR is mainly about the new structure, but also includes some new content. Happy to amend the new content if needed, or to address any feedback in follow-up issues (to avoid blocking the merge of this PR)
- The preview can be visualized at: https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/
- It contains a dark/light theme, courtesy of an updated version of the PyData Sphinx Theme
- We have enabled a few Sphinx directives for things like syntax highlighting and graphviz diagrams
- To build locally, create a venv, install dependencies, and then `make watch` to browse the content at `localhost:8000`
- I've updated the PR top-level comment with the "Closes ####" to link with issues that I think this PR will close once it is merged. It partially addresses other issues as well :tada: :partying_face: So we might have more issues being fixed after this one is merged, pending just some minor work.

And big thank you for the help & patience with this PR so far :bow: 

-Bruno

mr-c commented:
Broken links:
1. https://www.commonwl.org/user_guide/CODE_OF_CONDUCT.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/CODE_OF_CONDUCT.html

mr-c commented:
With regards to #141 ; I don't think there should be a distinction between intro topics and advanced topics

kinow commented:
>Broken links:
>
>    https://www.commonwl.org/user_guide/CODE_OF_CONDUCT.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/CODE_OF_CONDUCT.html (404)

Done! I added a redirect to <https://github.com/common-workflow-language/user_guide/blob/main/CODE_OF_CONDUCT.md>.

The previous user guide linked to an HTML page, with the GitHub `CODE_OF_CONDUCT.md` file rendered. But for citation instead of showing a rendered HTML, is had a link to the GitHub `CITATION` file.

@mr-c I made both point to the GitHub markdown file. Let me know if instead we should maintain what we had before, or maybe have both `CODE_OF_CONDUCT.md` and `CITATION` rendered directly in the site?

>    https://www.commonwl.org/user_guide/setup.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/setup.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/prerequisites.html (404)

Fixed, now pointing to `introduction/prerequisites.html`.

>    https://www.commonwl.org/user_guide/extras.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/extras.html (404)

I added a redirect to `/index.html`. We don't have the miscellaneous page any more, the YAML and the Recommended Practices are normal sections in the user guide. An alternative, I think, would be to add an orphan page with this content if it sounds useful to users?

>    https://www.commonwl.org/user_guide/misc/index.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/how-tos.html (404)

Moved to `/faq.html` where the old miscellaneous is now located at.

>    https://www.commonwl.org/user_guide/yaml/index.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/yaml/index.html (404)

Good catch! Fixed adding redirect `'yaml/index.md': '../essential-topics/yaml-guide.html',`

>    https://www.commonwl.org/user_guide/LICENSE.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/LICENSE.html (404)

Fixed!

![image](https://user-images.githubusercontent.com/304786/185258208-6e88cd16-a912-43f6-876b-b47e71596275.png)

>    https://www.commonwl.org/user_guide/01-introduction/index.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/quick-start.html (404)
>    https://www.commonwl.org/user_guide/02-1st-example/index.html → https://common-workflow-languageuser-guide--226.org.readthedocs.build/en/226/quick-start.html (404)

Both fixed now.

Thanks!

kinow commented:
>With regards to https://github.com/common-workflow-language/user_guide/issues/141 ; I don't think there should be a distinction between intro topics and advanced topics

@mr-c could you suggest how the topics must appear? Would it be something like:

```
1. Introduction
  1.1. Quick Start
  1.2. Prerequisites
  1.3. Basic Concepts
2. YAML Guide
3. Command Line TOol
4. Expressions
5. Inputs
6. Outputs
7. Custom Types
8. Expression Tool
9. Workflows
10. Operations
11. Environment Variables
12. Using Containers
13. Creating files at runtime
14. Staging input files
15. Best practices
16. File formats
17. Metadata and authorship
18. Specifying software requirements
19. Tutorials
20. FAQ
21. Licenses
```

? Note that this would make the top menu to have all these items, causing the top menu to have multiple lines, since Sphinx creates the menus at the left sidebar and at the top automatically from the table of contents. Maybe we could move everything under... Reference Material, or another section?

kinow commented:
Regarding #217 , we now have a link to the main website at the first paragraph, where it says "using the Common Workflow Language (CWL)", the Common Workflow Language text is now a link pointing to <https://www.commonwl.org/>.

Also redacted the text to say "open standards", and removed the part about ongoing work for new versions of the specification. However, I left the version used in the user guide as I thought that would be helpful information for readers (happy to amend if others prefer that version removed)? Here's a screen shot of the last commit.

![image](https://user-images.githubusercontent.com/304786/185259503-fd9259b5-dedc-40cd-8f6a-115af9c087e6.png)

kinow commented:
Have also confirmed we can redirect to sub-sections. @mr-c should we delete the old pages (i.e. `src/01-introduction/index.md`, `src/02-1st-example/index.md`, etc)? I realized these are not even rendered anymore since we have redirects in place now.

kinow commented:
Moved parameter references, that was in `outputs.md` to its own page, between Inputs and Outputs.

![image](https://user-images.githubusercontent.com/304786/185259919-d2dfc860-29c2-4885-83ca-65fb171363b7.png)

kinow commented:
Renamed Workflows / scattering workflows to Workflows / scattering steps

![image](https://user-images.githubusercontent.com/304786/185264461-9674f712-54b0-46a3-98a7-e5e79d5bb923.png)

Created follow-up https://github.com/common-workflow-language/user_guide/issues/232 for a better description of scattering workflows and scattering steps.

kinow commented:
Moved Additional Arguments and Parameters to its own section too, between Inputs and Outputs.

![image](https://user-images.githubusercontent.com/304786/185268402-c46d2423-a6d2-44c6-9d1f-4a62b374dd9d.png)

kinow commented:
Added a note to the `README.md` about the style guide, on specification vs. standard, and about not breaking links. Added building instructions to `CONTRIBUTING.md`.

mr-c commented:
> >With regards to https://github.com/common-workflow-language/user_guide/issues/141 ; I don't think there should be a distinction between intro topics and advanced topics


> 
> ? Note that this would make the top menu to have all these items, causing the top menu to have multiple lines, since Sphinx creates the menus at the left sidebar and at the top automatically from the table of contents. Maybe we could move everything under... Reference Material, or another section?
> 
> When creating this structure I looked at MDN (JS [Guides](https://developer.mozilla.org/en-US/docs/Learn/JavaScript) & [References](https://developer.mozilla.org/en-US/docs/Web/JavaScript) sections are sub-divided in basic/intermediate/advanced levels), [VueJS](https://vuejs.org/guide/introduction.html) (Essentials where you find most things to write Vue code, and then other sections for Components In-Depth, Best Practices, etc) 
> 

Hmm.. one of the motivations was to have all the topics listed on the left, so people don't have to switch between introductory and advanced. Can that be configured? 

Alternatively you can put all the intro and advanced sections under a single "topics" entry

kinow commented:
> > > With regards to #141 ; I don't think there should be a distinction between intro topics and advanced topics
> 
> > ? Note that this would make the top menu to have all these items, causing the top menu to have multiple lines, since Sphinx creates the menus at the left sidebar and at the top automatically from the table of contents. Maybe we could move everything under... Reference Material, or another section?
> > When creating this structure I looked at MDN (JS [Guides](https://developer.mozilla.org/en-US/docs/Learn/JavaScript) & [References](https://developer.mozilla.org/en-US/docs/Web/JavaScript) sections are sub-divided in basic/intermediate/advanced levels), [VueJS](https://vuejs.org/guide/introduction.html) (Essentials where you find most things to write Vue code, and then other sections for Components In-Depth, Best Practices, etc)
> 
> Hmm.. one of the motivations was to have all the topics listed on the left, so people don't have to switch between introductory and advanced. Can that be configured?
> 
> Alternatively you can put all the intro and advanced sections under a single "topics" entry

Ah, good suggestion. Let me try it locally and I'll either push it here for preview, or attach some screenshots. Thanks @mr-c!

kinow commented:
@mr-c last commit changes the outline to have a single "topics" section, which keeps the top menu functional. I simply merged by putting all the links from the old “Essential Topics” first, followed by the entries from “Advanced Topics”.

mr-c reviewed (approved):
If there are no objections, I would like to merge & deploy this, then work incrementally in smaller PRs to add more content & polish

kinow commented:
> If there are no objections, I would like to merge & deploy this, then work incrementally in smaller PRs to add more content & polish

/cc @tetron

tetron commented:
I approve this but I don't seem to be assigned as a reviewer so I can't formally hit the approve button?

mr-c commented:
Congratulations, thank you @kinow and reviewers!

kinow commented:
Thank you everybody for the help & patience! Feel free to send feedback as comments, messages, or follow-up issues :+1:
